File: dotnet/GHIElectronics.DUELink/Uart.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {

        public class UartController {

            SerialInterface serialPort;
            StreamController stream;

            public UartController(SerialInterface serialPort, StreamController stream) {
                this.serialPort = serialPort;
                this.stream = stream;
            }


            public bool Configuration(int baurdate, int rx_buffer_size) {

                var cmd = string.Format("SerCfg({0}, {1})", baurdate, rx_buffer_size);

                this.serialPort.WriteCommand(cmd);

                var ret = this.serialPort.ReadResponse();

                return ret.success;
            }

            public bool WriteByte(byte data) {
                var cmd = string.Format("SerWr({0})", data);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public int WriteBytes(byte[] data) {

                var cmd = $"dim b9[{data.Length}]";
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                var written = this.stream.WriteBytes("b9", data);

                cmd = string.Format("SerWrs(b9)");
                this.serialPort.WriteCommand(cmd);
                var ret = this.serialPort.ReadResponse();

                if (ret.success)
                    return written;

                return 0;
            }

            
            public byte ReadByte() {                

                this.serialPort.WriteCommand("SerRd()");

                var res = this.serialPort.ReadResponse();

                if (res.success) {
                    try {
                        var data = (byte)int.Parse(res.response);
                        return data;
                    }
                    catch {
                        
                    }

                }

                return 0;
            }

            public int ReadBytes(byte[] data, TimeSpan timeout) {

                var cmd = $"dim b9[{data.Length}]";
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();


                cmd = string.Format("SerRds(b9, {0})", timeout.TotalMilliseconds);
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                var ret = this.stream.ReadBytes("b9", data);

                return ret;
            }

            public int BytesToRead() {
                this.serialPort.WriteCommand("SerB2R()");

                var res = this.serialPort.ReadResponse();

                if (res.success) {
                    try {
                        var ready = int.Parse(res.response);
                        return ready;
                    }
                    catch {

                    }

                }
                return 0;
            }

            public bool Discard() {
                this.serialPort.WriteCommand("SerDisc()");

                var res = this.serialPort.ReadResponse();

                return res.success;
            }



        }
    }
}

File: python/DUELink/Uart.py
from DUELink.SerialInterface import SerialInterface
from DUELink.Stream import StreamController

class UartController:
    def __init__(self, transport:SerialInterface, stream:StreamController):
        self.transport = transport
        self.stream = stream

    def Configuration(self, baudrate: int, rx_buf_size: int)->bool:
        cmd = "SerCfg({0}, {1})".format(baudrate, rx_buf_size)
        self.serialport.WriteCommand(cmd)
        ret = self.serialport.ReadResponse()
        return ret.success

    def WriteByte(self, data: int)->bool:
        cmd = "SerWr({})".format(data)
        self.serialport.WriteCommand(cmd)
        ret = self.serialport.ReadResponse()
        return ret.success        
    
    def WriteBytes(self, data: bytes)->int:
        count = len(data)
        cmd = f"dim b9[{count}]"
        self.transport.WriteCommand(cmd)
        self.transport.ReadResponse()

        written = self.stream.WriteBytes("b9", data)

        self.serialport.WriteCommand("SerWrs(b9)")
        ret = self.transport.ReadResponse()

        if (ret.success):
            return written

        return 0
    
    def ReadByte(self):        
        self.serialport.WriteCommand("SerRd()")
        res = self.serialport.ReadResponse()
        if res.success:
            try:
                data = int(res.response)
                return data
            except:
                pass
        return 0
    
    def ReadBytes(self, data: bytearray, timeout_ms: int)->int:
        count = len(data)
        cmd = f"dim b9[{count}]"
        self.transport.WriteCommand(cmd)
        self.transport.ReadResponse()

        cmd = f"SerRds(b9, {timeout_ms})"
        self.serialport.WriteCommand(cmd)
        ret = self.transport.ReadResponse()

        read = self.stream.ReadBytes("b9",data )

        if (ret.success):
            return read

        return 0

    def BytesToRead(self)->int:        
        self.serialport.WriteCommand("SerB2R()")
        ret = self.serialport.ReadResponse()
        if ret.success:
            try:
                ready = int(ret.response)
                return ready
            except:
                pass
        return 0
    
    def Discard(self)->bool:        
        self.serialport.WriteCommand("SerDisc()")
        ret = self.serialport.ReadResponse()
        
        return ret.success
    

    
